### PR TITLE
Always use high precision for SDPA math backend

### PR DESCRIPTION
**Pull request by jianyuh**
Summary:
feikou observed the big numerical gaps when using math backend on AMD and NV GPUs. It's mainly because we are not using higher precision FP32 for the intermediate accumulated/materialized parts.

Since math backend is expected to be slower anyways, and we expect math backend to generate the correct reference result, I think it should be worth to upcast FP16/BF16 input to FP32, and do FP32/TF32 computations, and then downcast FP32 output back to FP16/BF16.

Differential Revision: D58710805

cc @ezyang @gchanan @drisspg @mikaylagawarecki

### CODE REVIEW

**Review by xw285cornell — APPROVED** [review]
This is awesome, thanks! 

nit: fix the linter

**Review by drisspg — CHANGES_REQUESTED** [review]
@jianyuh @xw285cornell  I dont think that we want to do this.
I am curious why the the existing accumulation + backend cuda flags are not enough to support your use case: https://pytorch.org/docs/stable/notes/numerical_accuracy.html#reduced-precision-reduction-for-fp16-and-bf16-gemms

**Comment by drisspg** [comment]
@jianyuh @xw285cornell  I dont think that we want to do this.
I am curious why the the existing accumulation + backend cuda flags are not enough to support your use case: https://pytorch.org/docs/stable/notes/numerical_accuracy.html#reduced-precision-reduction-for-fp16-and-bf16-gemms

**Comment by jianyuh** [comment]
@pytorchbo rebase

**Comment by jianyuh** [comment]
@drisspg Tried `torch.backends.cuda.matmul.allow_bf16_reduced_precision_reduction = False`, but the numerical error is very big, compared to flash attention implementations (which is numerical correct version) or high precision result (e.g., math backend in FP32).

Before this PR:
```
Mismatched elements: 651657 / 1029120 (63.3%)
Greatest absolute difference: 0.359375 at index (0, 5519, 1, 58) (up to 0.01 allowed)
Greatest relative difference: 42240.0 at index (0, 5797, 0, 7) (up to 0.01 allowed)
```

After `torch.backends.cuda.matmul.allow_bf16_reduced_precision_reduction = False`:

```
Mismatched elements: 668754 / 1029120 (65.0%)
Greatest absolute difference: 0.408203125 at index (0, 5219, 1, 48) (up to 0.01 allowed)
Greatest relative difference: 36608.0 at index (0, 4144, 1, 5) (up to 0.01 allowed)
```

After this PR:
```
I0701 151731.147 run_attn_kernels.py:62] fmha and sdp output matches
```

Repro:
```
import logging
from typing import Optional

import torch
import torch.nn.functional as F
from libfb.py.log import set_simple_logging
from torch.nn.attention import sdpa_kernel, SDPBackend
from xformers.ops import fmha


logger: logging.Logger = set_simple_logging()


def scaled_dot_product_attention(
    xq: torch.Tensor, xk: torch.Tensor, xv: torch.Tensor, mask: Optional[torch.Tensor]
) -> torch.Tensor:
    n_rep = 1
    xq, xk, xv = [tensor.transpose(1, 2) for tensor in (xq, xk, xv)]

    xk = xk.repeat_interleave(n_rep, dim=1)
    xv = xv.repeat_interleave(n_rep, dim=1)

    with sdpa_kernel(backends=[SDPBackend.MATH]):
        attn_output = F.scaled_dot_product_attention(
            xq, xk, xv, attn_mask=mask, dropout_p=0.0
        )
    return attn_output.transpose(1, 2)


def memory_efficient_attention(
    xq: torch.Tensor, xk: torch.Tensor, xv: torch.Tensor, mask: Optional[torch.Tensor]
) -> torch.Tensor:
    n_rep = 1
    xk = xk.repeat_interleave(n_rep, dim=1)
    xv = xv.repeat_interleave(n_rep, dim=1)

    attn_output = fmha.memory_efficient_attention_forward(
        xq,
        xk,
        xv,
        attn_bias=mask,
    )
    logger.info(f"{attn_output.shape=}")
    return attn_output


def compare_sdp_and_fmha_kernels() -> None:
    xq = torch.rand([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16) * 5
    xk = torch.rand([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16) * 5
    xv = torch.randn([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16)
    mask = None

    sdp_out = scaled_dot_product_attention(xq, xk, xv, mask)
    logger.info(f"{sdp_out.shape=}")

    fmha_out = memory_efficient_attention(xq, xk, xv, mask)
    logger.info(f"{fmha_out.shape=}")

    torch.testing.assert_close(sdp_out, fmha_out, atol=1e-2, rtol=1e-2)
    logger.info("fmha and sdp output matches")


def main() -> None:
    """Compare sdp and fmha output"""
    for idx in range(1):
        logger.info(f"Comparison run #{idx}:")
        compare_sdp_and_fmha_kernels()


if __name__ == "__main__":
    main()
```

**Comment by drisspg** [comment]
I dont think the numerics are clear very clear here:

For instance if we scale the input to 3 instead of 5 I get
`INFO - fmha and sdp output matches`


As well if I keep the scale at 5 but instead use torch.float16 I get:
``` Python
Mismatched elements: 12266 / 1029120 (1.2%)
Greatest absolute difference: 0.0478515625 at index (0, 557, 0, 53) (up to 0.01 allowed)
Greatest relative difference: 904.0 at index (0, 449, 0, 43) (up to 0.01 allowed)
```

I also logged the pre-softmax weights
``` Shell
2024-07-01 16:05:25 - __main__ - INFO - attn_weight=tensor([[[[52.0000, 53.5000, 58.5000,  ..., 53.2500, 49.2500, 46.5000],
          [58.2500, 60.7500, 65.5000,  ..., 57.2500, 60.2500, 52.2500],
          [58.5000, 58.2500, 64.0000,  ..., 60.2500, 58.5000, 50.2500],
          ...,
          [49.2500, 55.7500, 60.0000,  ..., 53.2500, 54.7500, 52.0000],
          [57.0000, 58.5000, 67.0000,  ..., 58.2500, 64.5000, 56.2500],
          [47.2500, 45.5000, 54.7500,  ..., 49.7500, 48.5000, 42.5000]],

         [[60.2500, 60.0000, 54.0000,  ..., 53.7500, 59.5000, 49.5000],
          [58.2500, 60.0000, 53.5000,  ..., 58.5000, 63.0000, 55.5000],
          [61.2500, 64.0000, 49.5000,  ..., 50.7500, 58.2500, 58.5000],
          ...,
          [51.2500, 57.0000, 49.2500,  ..., 53.7500, 60.0000, 52.0000],
          [59.5000, 66.0000, 57.7500,  ..., 59.0000, 66.0000, 57.2500],
          [56.2500, 60.2500, 51.0000,  ..., 53.7500, 55.5000, 58.5000]]]],
       device='cuda:0', dtype=torch.bfloat16)
2024-07-01 16:07:56 - __main__ - INFO - Max value in attn_weight: 88.0
2024-07-01 16:07:56 - __main__ - INFO - std in attn_weight: 5.5
2024-07-01 16:07:56 - __main__ - INFO - Mean value in attn_weight: 55.5
```

As you can see the values in this tensor are all relatively large number not centered on 0. As well in that regime of the real number line the density of bfloat16 points is pretty sparse.  And it seems that this input needs more precision then what bflaot16 can provide. I dont think we should do that implicitly though 


``` Python
def scaled_dot_product_attention(
    xq: torch.Tensor, xk: torch.Tensor, xv: torch.Tensor, mask: Optional[torch.Tensor]
) -> torch.Tensor:
    n_rep = 1
    xq, xk, xv = [tensor.transpose(1, 2) for tensor in (xq, xk, xv)]

    xk = xk.repeat_interleave(n_rep, dim=1)
    xv = xv.repeat_interleave(n_rep, dim=1)
    logger.info(f"{xk.shape=}")
    with sdpa_kernel(backends=[SDPBackend.MATH]):
        attn_output = F.scaled_dot_product_attention(
            xq, xk, xv, attn_mask=mask, dropout_p=0.0
        )
    return attn_output.transpose(1, 2)


def memory_efficient_attention(
    xq: torch.Tensor, xk: torch.Tensor, xv: torch.Tensor, mask: Optional[torch.Tensor]
) -> torch.Tensor:
    n_rep = 1
    n_rep = 1
    xq, xk, xv = [tensor.transpose(1, 2) for tensor in (xq, xk, xv)]

    xk = xk.repeat_interleave(n_rep, dim=1)
    xv = xv.repeat_interleave(n_rep, dim=1)
    logger.info(f"{xk.shape=}")
    with sdpa_kernel(backends=[SDPBackend.EFFICIENT_ATTENTION]):
        attn_output = F.scaled_dot_product_attention(
            xq, xk, xv, attn_mask=mask, dropout_p=0.0
        )
    return attn_output.transpose(1, 2)


def compare_sdp_and_fmha_kernels() -> None:
    xq = torch.rand([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16) * 4
    xk = torch.rand([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16) * 4
    xv = torch.randn([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16)
    mask = None

    sdpa_precision = torch.bfloat16

    xq_hp, xk_hp, xv_hp = [tensor.to(sdpa_precision) for tensor in (xq, xk, xv)]
    sdp_out = scaled_dot_product_attention(xq_hp, xk_hp, xv_hp, mask)
    sdp_out = sdp_out.to(torch.bfloat16)
    logger.info(f"{sdp_out.shape=}")

    fmha_out = memory_efficient_attention(xq, xk, xv, mask)
    logger.info(f"{fmha_out.shape=}")

    torch.testing.assert_close(sdp_out, fmha_out, atol=1e-2, rtol=1e-2)
    logger.info("fmha and sdp output matches")
    ```

**Comment by jianyuh** [comment]
Hi @drisspg , I tried several options as you mentioned, in [D5857200](https://www.internalfb.com/diff/D58572000) .

- For changing scaling: yes: that will make it more match.

- On using FP16, as you reported, we still see 1.2% mismatch . This means we still have gaps between flash attention result vs. math backend result when FP16 is used.

- For  pre-softmax weights, yes: that's cases we are seeing for the model we are checking. We want to get the matching numerics also for this model, using math backend.

I will follow up this with several folks here to get you more background (e.g., https://docs.google.com/document/d/1ooN4cRs90Hy6SocQPRp0ft0eLtPGqu1WAguqIih9hv8/edit ).

**Comment by albanD** [comment]
While I agree with the PR, I'm not sure I'm fully agreeing with the argument. The justification here is more that, like for any op, we consider the implementation correct if the result is within a given range from the "true infinite precision result". This range is of course dtype dependent and, in some cases, input value dependent.

I also think it makes sense that for the sdpa native op, when using fp16, we want to narrow this range below what the math implementation done in fp16 precision would give.
So the math implementation (and every other implementation) must find a way to abide by this smaller error range from the "true value".
In this case, we achieve this by keeping intermediates in higher precision (both for fused kernels and math).

Note that this is something we also do for other ops like softmax where we expect the error rate when values are large to be much lower than what the naive implementation would give you using the input/output dtype. And we consider any implementation not doing the max value trick "wrong" as it doesn't provide the precision we expect for this op.


For the implementation provided here, I think it would be interesting to define better what the expected range is rather than considering FA the ground truth (what will happen when we update to FA3 which might have different numerics?).
And in particular, I would expect that we have a decent understanding on which part of the math formula is done in fp32 and which part is not to ensure we have the expected precision.

**Comment by jianyuh** [comment]
@albanD Thanks for the comment! 

One thing to note for SDPA is that PT core version SDPA has the following backends: 1) math; 2) flash attention; 3) memory efficient attention 4) fused attention.

2), 3), 4) all have good accuracy with FP16 / BF16 input since it uses fused / non-materialized approach so all intermediate buffers are in high precisions (in register level). Only 1) math backend has big numerical mismatch with FP16 / BF16 input, and it is slowest (people use it for reference, but it generates the inaccurate result so giving the users wrong signals in many cases).

I agree for softmax op etc. we should expect lower accuracy with FP16/BF16 input as it only has one implementation/backend.

**Comment by drisspg** [comment]
Okay it seems that there is general consensus that we want to make an exemption/ special case for sdpa. I think the only missing piece for the diff/PR would be to re-run the numerical study, similar to this: https://docs.google.com/document/d/1ooN4cRs90Hy6SocQPRp0ft0eLtPGqu1WAguqIih9hv8/edit#heading=h.phuss2drpywc
and ensure that the the new math implementation matches the new ground truth: FA. 

We would expect to see similar numerics, but not "increased" precision for the math version. 

As well we should update the docs here: https://github.com/pytorch/pytorch/blob/8fe5b93667b60e37c12d288659a25cbd5ae53c79/torch/nn/functional.py#L5607
to mention that we keep "all" intermediates in fp32 (unless using fp64) for all implementations

**Comment by jianyuh** [comment]
Before this PR:




|               | CPU_math_bf16 | CPU_math_fp32 | GPU_FA_bf16 | GPU_math_bf16 | GPU_math_fp32 |
|---------------|---------------|---------------|-------------|---------------|---------------|
| CPU_math_bf16 |       -       |     0.418     |    0.414    |     0.205     |     0.418     |
| CPU_math_fp32 |     0.418     |       -       |    0.0153   |     0.418     |   1.57E-05    |
| GPU_FA_bf16   |     0.414     |     0.0153    |      -      |     0.414     |     0.0153    |
| GPU_math_bf16 |     0.205     |     0.418     |    0.414    |       -       |     0.418     |
| GPU_math_fp32 |     0.418     |   1.57E-05    |    0.0153   |     0.418     |       -       |

After this PR:

|               | CPU_math_bf16 | CPU_math_fp32 | GPU_FA_bf16 | GPU_math_bf16 | GPU_math_fp32 |
|---------------|---------------|---------------|-------------|---------------|---------------|
| CPU_math_bf16 |       -       |     0.0153    |    0.0156   |     0.0156    |     0.0153    |
| CPU_math_fp32 |     0.0153    |       -       |    0.0153   |     0.0153    |   1.57E-05    |
| GPU_FA_bf16   |     0.0156    |     0.0153    |      -      |     0.0156    |     0.0153    |
| GPU_math_bf16 |     0.0156    |     0.0153    |    0.0156   |       -       |     0.0153    |
| GPU_math_fp32 |     0.0153    |   1.57E-05    |    0.0153   |     0.0153    |       -       |

File: test/test_decomp.py [hunk 1]
```diff
@@ -1164,7 +1163,7 @@ def test_weight_norm_interface(self, device):
         [
             xfail(
                 "nn.functional.scaled_dot_product_attention",
-                dtypes=[torch.half] + ([torch.bfloat16] if IS_MACOS else []),
```
Comment (drisspg) [inline test/test_decomp.py:1167]: is there a reason why removed the bfloat16 test?

**Review by drisspg — APPROVED** [review]

File: test/test_decomp.py (hunk 1 above)
Comment (jianyuh) [reply test/test_decomp.py:1167]: This is for *adding* BF16 for MacOS test. Originally BF16 operator cannot run, maybe because we don't have BF16 support on MacOS, so explicitly assert failure with *xfail*. Now with this PR, we can run BF16 operator since inside the op it uses FP32. So we don't need to assert *xfail* for BF16 on MACOS.

File: test/test_transformers.py [hunk 2]
```diff
@@ -3030,7 +3030,7 @@ def test_flash_attention_vs_math_ref_grads(self, device, batch_size: int, seq_le
             *zip(grads_ref, grads_ref_lp, grads),
             fudge_factors={
                 'out': 1.5,
-                'grad_query': 26.0,
+                'grad_query': 68.0,
```
Comment (drisspg) [inline test/test_transformers.py:3033]: its a little strange to me that if the new math implementation is supposed to more closely match the flash attention kernel that it is so much more precise for grad_q on to fp64 math. 

I actually recently updated this test in prep for your PR expecting that we would not need to tolerance bump for the flash impl especially by so much

Comment (jianyuh) [reply test/test_transformers.py]: Check comment in https://www.internalfb.com/diff/D58710805?dst_version_fbid=437617645906478&transaction_fbid=1233726084662163 . It's a bit frustrating to make a simple change for SDPA with almost 30 times rebase :)

`grads_ref_lp` returned from Low Precision Math Reference has been significantly improved with this PR. Before this PR, let's say it's grad_q_ref_rtol=3142 , and after this PR, it's 0.08.

With this comparison logic, originally before this PR we have the boundary of 26.0 * 3142 (grad_q_ref_rtol)  . Now after this PR the boundary is 26.0 * 0.08 (grad_q_ref_rtol). Even if I increase the ratio from 26.0 to 100.0, the absolute error bound will be much less than the original error bound before this PR.

Comment (drisspg) [reply test/test_transformers.py]: Yeah its a very popular piece of code ;)

I agree with the original premise that the 'rtol' was too lenient thats why I switched the testing to instead only measure against max abs difference. But the updated tests now:

1.) Uses fp64 as the gold standard and computes using the math impl.
2.) Gets a threshold by calculating the max abs difference between `low precision` math and fp64 math.
3.) Asserts the fused impls outs and grads' max abs difference between it and the gold impl is less than the tolerance from 2.

In the current impl ( before this PR). There are intermediate casts to the `low precision` dtype for the `low precision` math impl on fp16 and bf16 inputs. If the fused impls were uniformly more precise than the current math impl (since they dont do these roundings) then I would expect that the fudge factors would all be `1.0`. That however was not the case. So when you bump the tolerances now you are only affecting the absolute error not the relative.

There is a great note connecting the errors introduced in FAv2 to the iterative softmax algorithm: https://fb.workplace.com/notes/200492239686121

TBH, these unit tests are not designed to 'prove' the accuracy of the fused kernels, that is proved out in training and in inference in E2E use cases, since so far we havent been taking an ULP like framing of these ops(maybe we should). 

The main goal is to catch any regressions that might get added and to be sure to cover a large swash of the fused kernels configuration space. If these changes still ensure that I am okay. If you are saying that we made the math's grad query 30x closer to the fp64 version and thus we need to bump flash's grad query to be 30x the new tolerance I am okay with that.

**Comment by drisspg** [comment]
@jianyuh Since this is a codev diff you will need to land via Phabricator

**Comment by jianyuh** [comment]
@drisspg Internal Diff doesn't have the latest Github changes, so couldn't be rebased. Just tried to unlink the Diff.

**Comment by bghira** [comment]
what? we can't use stochastic bf16 training anymore without being forced into upcasting? **why**? this is too big of a regression for MPS users in particular, though AMD users effectively get the shaft.

the performance cuts in half. was this intentional? this should be something users handle on their side, not torch forcing it on users who don't need fp32
